### PR TITLE
Verify backend is up before running frontend in start scripts

### PR DESCRIPTION
**Pull request by m2web**
### Motivation and Context
I noted that when running the Start.ps1 the backend would not be running via `Start-Process pwsh -ArgumentList "-command $BackendScript"` before the frontend command `& $FrontendScript`. I added a condition to check that the backend is running before starting the frontend process.

### Description
I added a condition to check that the backend is running before starting the frontend process.

Extension of: https://github.com/microsoft/chat-copilot/pull/310 with changes:
Per @teresaqhoang I wanted to set a retry count ceiling in the `/scripts/Start.*` scripts files when checking if the backend service was up and working and set a limit of retries "in case the backend server goes dark for any reason." It's a good idea. 😊 

Note commit #f0d049b. Thanks.

### Description

Set a maximum number of retries on the backend service call.

### Contribution Checklist

- [x ] The code builds clean without any errors or warnings
- [x ] The PR follows the [Contribution Guidelines](https://github.com/microsoft/chat-copilot/blob/main/CONTRIBUTING.md) and the [pre-submission formatting script](https://github.com/microsoft/chat-copilot/blob/main/CONTRIBUTING.md#development-scripts) raises no violations
- [ ] All unit tests pass, and I have added new tests where possible
- [x ] I didn't break anyone :smile:

### CODE REVIEW

File: scripts/start.sh [hunk 1]
```diff
@@ -7,8 +7,18 @@ set -e
...
 
+# get the port from the REACT_APP_BACKEND_URI env variable
+envContent=$(grep -v '^#' ../webapp/.env | xargs)
+backendPort=$(echo $envContent | sed -n 's/.*:\([0-9]*\).*/\1/p')
+
 # Start backend (in background)
 ./start-backend.sh &
 
+# check that the backend is running and keep checking until it is
+while ! nc -z localhost $backendPort; do
```
Comment (teresaqhoang) [inline scripts/start.sh:18]: Please add the retry logic to this script too

Comment (m2web) [reply scripts/start.sh]: If you look at the start.sh file diff I did implement the retry logic.  Am I missing something?

Comment (glahaye) [reply scripts/start.sh]: Looks better now except you've removed the .gitignore from the repo!

Please push it back in and this PR should be good to go.

Comment (m2web) [reply scripts/start.sh]: Yep, sorry about that.

File: .gitignore [hunk 2]
```diff
@@ -489,4 +489,7 @@ webapi/CopilotChatWebApi.sln
 /deploy/
 
 # Tesseract OCR language data files
-*.traineddata
\ No newline at end of file
+*.traineddata
+
+# Auto-generated JSON file from Visual Studio
+webapi/appsettings.Development.json
```
Comment (glahaye) [inline .gitignore:495]: You don't need addition,  Line 417 already takes care of this file (verified on my machine)

File: scripts/Start.ps1 [hunk 3]
```diff
@@ -8,6 +8,41 @@ $FrontendScript = Join-Path "$PSScriptRoot" 'Start-Frontend.ps1'
...
+$backendRunning = $false
 
-# Start frontend (in current PS process)
-& $FrontendScript
+# get the port from the REACT_APP_BACKEND_URI env variable
+$envFilePath = Join-Path $PSScriptRoot '..\webapp\.env'
+$envContent = Get-Content -Path $envFilePath
+$port = [regex]::Match($envContent, ':(\d+)/').Groups[1].Value
+
+# set max retries to 5
```
Comment (glahaye) [inline scripts/Start.ps1:19]: This comment should be removed: it just repeats what's self evident from the code

File: scripts/Start.ps1 [hunk 4]
```diff
@@ -8,6 +8,41 @@ $FrontendScript = Join-Path "$PSScriptRoot" 'Start-Frontend.ps1'
...
+$maxRetries = 5
+
+# set retry count to 0
```
Comment (glahaye) [inline scripts/Start.ps1:22]: This comment should be removed: it just repeats what's self evident from the code

File: scripts/start.sh [hunk 5]
```diff
@@ -7,8 +7,45 @@ set -e
...
+# get the port from the REACT_APP_BACKEND_URI env variable
+envContent=$(grep -v '^#' ../webapp/.env | xargs)
+backendPort=$(echo $envContent | sed -n 's/.*:\([0-9]*\).*/\1/p')
+
 # Start backend (in background)
 ./start-backend.sh &
 
-# Start frontend
-./start-frontend.sh
+# set max retries to 5
```
Comment (glahaye) [inline scripts/start.sh:17]: This comment should be removed: it just repeats what's self evident from the code

File: scripts/start.sh [hunk 6]
```diff
@@ -7,8 +7,45 @@ set -e
...
+maxRetries=5
+
+# set retry count to 0
```
Comment (glahaye) [inline scripts/start.sh:20]: This comment should be removed: it just repeats what's self evident from the code

File: scripts/Start.ps1 [hunk 7]
```diff
@@ -8,6 +8,39 @@ $FrontendScript = Join-Path "$PSScriptRoot" 'Start-Frontend.ps1'
...
+# check if the backend is running before proceeding
+$backendRunning = $false
 
-# Start frontend (in current PS process)
-& $FrontendScript
+# get the port from the REACT_APP_BACKEND_URI env variable
+$envFilePath = Join-Path $PSScriptRoot '..\webapp\.env'
+$envContent = Get-Content -Path $envFilePath
+$port = [regex]::Match($envContent, ':(\d+)/').Groups[1].Value
+$maxRetries = 5
```
Comment (glahaye) [inline scripts/Start.ps1:18]: Inserting a space makes this cleaner

```suggestion

$maxRetries = 5
```

File: scripts/Start.ps1 [hunk 8]
```diff
@@ -8,6 +8,40 @@ $FrontendScript = Join-Path "$PSScriptRoot" 'Start-Frontend.ps1'
...
-# Start frontend (in current PS process)
-& $FrontendScript
+# get the port from the REACT_APP_BACKEND_URI env variable
+$envFilePath = Join-Path $PSScriptRoot '..\webapp\.env'
+$envContent = Get-Content -Path $envFilePath
+$port = [regex]::Match($envContent, ':(\d+)/').Groups[1].Value
+
+$maxRetries = 5
+
+# set retry count to 0
```
Comment (glahaye) [inline scripts/Start.ps1:21]: Removing this comment:

```suggestion

```

File: scripts/Start.ps1 [hunk 9]
```diff
@@ -8,6 +8,38 @@ $FrontendScript = Join-Path "$PSScriptRoot" 'Start-Frontend.ps1'
...
+# get the port from the REACT_APP_BACKEND_URI env variable
+$envFilePath = Join-Path $PSScriptRoot '..\webapp\.env'
+$envContent = Get-Content -Path $envFilePath
+$port = [regex]::Match($envContent, ':(\d+)/').Groups[1].Value
+
+$maxRetries = 5
+$retryCount = 0
+
+# set the number of seconds to wait before retrying
+$retryWait = 5
```
Comment (glahaye) [inline scripts/Start.ps1:23]: Reformatting

```suggestion
$retryCount = 0
$retryWait = 5 # set the number of seconds to wait before retrying
```

File: .gitignore [hunk 10]
```diff
@@ -489,4 +489,4 @@ webapi/CopilotChatWebApi.sln
 /deploy/
 
 # Tesseract OCR language data files
-*.traineddata
\ No newline at end of file
+*.traineddata
```
Comment (glahaye) [inline .gitignore:492]: Reverting removal of line feed

```suggestion
*.traineddata

```

**Review by glahaye — APPROVED** [review]

File: scripts/Start.ps1 [hunk 11]
```diff
@@ -8,6 +8,36 @@ $FrontendScript = Join-Path "$PSScriptRoot" 'Start-Frontend.ps1'
...
+  $backendRunning = Test-NetConnection -ComputerName localhost -Port $port -InformationLevel Quiet
+  Start-Sleep -Seconds $retryWait
+}
+
+# if the backend is running, start the frontend
+if ($backendRunning -eq $true) {
+  # Start frontend (in current PS process)
+  & $FrontendScript
+}
+else {
```
Comment (TaoChenOSU) [inline scripts/Start.ps1:35]: nit
```suggestion
} else { 
```

File: scripts/Start.ps1 [hunk 12]
```diff
@@ -8,6 +8,36 @@ $FrontendScript = Join-Path "$PSScriptRoot" 'Start-Frontend.ps1'
...
+# if the backend is running, start the frontend
+if ($backendRunning -eq $true) {
+  # Start frontend (in current PS process)
+  & $FrontendScript
+}
+else { 
+  # otherwise, write to the console that the backend is not running and we have exceeded the number of retries and we are exiting
+  Write-Host "*************************************************"
+  Write-Host "Backend is not running and we have exceeded "
+  Write-Host "the number of retries."
```
Comment (TaoChenOSU) [inline scripts/Start.ps1:39]: nit:
```suggestion
  Write-Host "the maximum number of retries."
```

File: scripts/start.sh [hunk 13]
```diff
@@ -7,8 +7,40 @@ set -e
...
+backendPort=$(echo $envContent | sed -n 's/.*:\([0-9]*\).*/\1/p')
+
 # Start backend (in background)
 ./start-backend.sh &
 
-# Start frontend
-./start-frontend.sh
+maxRetries=5
+retryCount=0
+retryWait=5
```
Comment (TaoChenOSU) [inline scripts/start.sh:19]: nit: add the following comment like in the other script
```suggestion
retryWait=5  # set the number of seconds to wait before retrying
```

File: scripts/start.sh [hunk 14]
```diff
@@ -7,8 +7,40 @@ set -e
...
+  fi
+done
+
+# if we have exceeded the number of retries
+if [ $retryCount -eq $maxRetries ]
+then
+# write to the console that the backend is not running and we have exceeded the number of retries and we are exiting
+  echo "*************************************************"
+  echo "Backend is not running and we have exceeded "
+  echo "the number of retries."
```
Comment (TaoChenOSU) [inline scripts/start.sh:42]: nit:
```suggestion
  echo "the maximum number of retries."
```

File: scripts/Start.ps1 [hunk 15]
```diff
@@ -8,6 +8,36 @@ $FrontendScript = Join-Path "$PSScriptRoot" 'Start-Frontend.ps1'
...
+  $backendRunning = Test-NetConnection -ComputerName localhost -Port $port -InformationLevel Quiet
+  Start-Sleep -Seconds $retryWait
+}
+
+# if the backend is running, start the frontend
+if ($backendRunning -eq $true) {
+  # Start frontend (in current PS process)
+  & $FrontendScript
+}
+} else {
```
Comment (glahaye) [inline scripts/Start.ps1:35]:
```suggestion
} else { 
```

**Review by glahaye — APPROVED** [review]

**Review by TaoChenOSU — APPROVED** [review]

**Pull request merged**